notebooks/00_preprocess_data.ipynb

```diff
@@ -0,0 +1,494 @@
+    "\n",
+    "# Subset the dataframes to the relevant fields\n",
+    "plot_df = plot_df[[\"PLOT_ID\", \"PLT_CN\", \"ASSESSMENT_YEAR\", \"X\", \"Y\"]]\n",
+    "tree_df = tree_df[[\"TREE_ID\", \"TREE_CN\", \"PLOT_ID\", \"SPECIES_SYMBOL\", \"DBH_CM\", \"TPH\"]]\n",
+    "\n",
+    "# Show the first few rows of each dataframe\n",
+    "print(plot_df.head())\n",
+    "print(tree_df.head())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Step 5: Create a species by DBH-class matrix of basal area for GNN modeling"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The Gradient Nearest Neighbor (GNN) method needs both a species matrix and an environmental matrix (next step) to run.  Typically, we have used species by 25-cm diameter classes to create the species matrix.  We can create this matrix from `tree_df` by running the following code:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def create_species_matrix(tree_df: pd.DataFrame, plot_df: pd.DataFrame) -> pd.DataFrame:\n",
+    "    df = tree_df.copy()\n",
+    "\n",
+    "    # Add tree basal area and basal area per hectare columns\n",
+    "    df[\"BA_M2\"] = (df.DBH_CM ** 2) * 0.00007854\n",
+    "    df[\"BAPH\"] = df.BA_M2 * df.TPH\n",
+    "\n",
+    "    # Create a new column called DBH_GROUP which grounps the DBH_CM column\n",
+    "    # these increments [<25, 25-50, 50-75, 75-100, >100].  Call these classes\n",
+    "    # 1 - 5.\n",
+    "    dbh_bins = [0, 25, 50, 75, 100, 1000]\n",
+    "    df[\"DBH_GROUP\"] = pd.cut(df.DBH_CM, bins=dbh_bins, labels=[1, 2, 3, 4, 5])\n",
+    "\n",
+    "    # Combine the SPECIES_SYMBOL and DBH_GROUP fields into a new column\n",
+    "    # called SPECIES_SIZE_CLASS separated by an underscore\n",
+    "    df[\"SPECIES_SIZE_CLASS\"] = df[\"SPECIES_SYMBOL\"] + \"_\" + df[\"DBH_GROUP\"].astype(str)\n",
+    "\n",
+    "    # Group by this new column and sum the BAPH column\n",
+    "    grouped_df = df.groupby([\"PLOT_ID\", \"SPECIES_SIZE_CLASS\"]).BAPH.sum()\n",
+    "\n",
+    "    # Pivot the data into a matrix\n",
+    "    grouped_df = grouped_df.reset_index()\n",
+    "    grouped_df = grouped_df.pivot(\n",
+    "        index=\"PLOT_ID\", columns=\"SPECIES_SIZE_CLASS\", values=\"BAPH\"\n",
+    "    )\n",
+    "\n",
+    "    # Merge plot_df with the grouped_df to account for plots with no tree\n",
+    "    # records and fill the missing NA values with 0\n",
+    "    grouped_df = plot_df[[\"PLOT_ID\"]].merge(grouped_df, on=\"PLOT_ID\", how=\"left\")\n",
+    "    grouped_df = grouped_df.fillna(0)\n",
+    "\n",
+    "    # Get the sum of all but the PLOT_ID column to find plots with no tree records\n",
+    "    row_sums = grouped_df.iloc[:,1:].sum(axis=1)\n",
+    "\n",
+    "    # If row_sums is 0, set a NOTALY field to 0.001, otherwise set to 0\n",
```

Comment:
This might be opening a can of worms, but worth a quick explanation of why we need to do this?

Reply:
Not at all.  Good important information, without which it's a bit confusing what I'm doing with the `NOTALY` column.  I've just added a bit more explanation here. in the markdown above this cell.